var/spack/repos/builtin/packages/siesta/package.py

```diff
@@ -49,49 +98,209 @@ def flag_handler(self, name, flags):
-            sh = which("sh")
-            sh("build_all.sh")
+        if spec.satisfies("@:4.0.2"):
+            with working_dir("Obj_trans"):
+                make("transiesta", parallel=False)
+            with working_dir("Util"):
+                sh = which("sh")
+                sh("build_all.sh")
 
     def install(self, spec, prefix):
         mkdir(prefix.bin)
         with working_dir("Obj"):
             install("siesta", prefix.bin)
-        with working_dir("Obj_trans"):
-            install("transiesta", prefix.bin)
+        if spec.satisfies("@:4.0.2"):
+            with working_dir("Obj_trans"):
+                install("transiesta", prefix.bin)
         for root, _, files in os.walk("Util"):
             for fname in files:
                 fname = join_path(root, fname)
                 if os.access(fname, os.X_OK):
                     install(fname, prefix.bin)
+
+
+class CMakeBuilder(cmake.CMakeBuilder):
+    """Use the new CMake build system to build siesta@5.0.0:."""
+
+    def cmake_args(self):
+        spec = self.spec
+        args = []
+
+        args += ["-DBLAS_LIBRARIES={0}".format(self.spec["blas"].libs.link_flags)]
+        args += ["-DLAPACK_LIBRARIES={0}".format(self.spec["lapack"].libs.link_flags)]
+
+        if "+mpi" in spec:
+            args += [
+                "-DCMAKE_C_COMPILER=%s" % spec["mpi"].mpicc,
+                "-DCMAKE_CXX_COMPILER=%s" % spec["mpi"].mpicxx,
+                "-DCMAKE_Fortran_COMPILER=%s" % spec["mpi"].mpifc,
+            ]
+            args += ["-DSIESTA_WITH_MPI=ON"]
+            args += ["-DSCALAPACK_LIBRARY={0}".format(spec["scalapack"].libs.joined(";"))]
+
+        if "+openmp" in spec:
+            args += ["-DSIESTA_WITH_OPENMP=ON"]
+            if "+cray" in spec:
+                args += ["-DFortran_FLAGS=-fopenmp"]
+
+        if "build_type=Debug" in spec:
+            args += [
+                "-DFortran_FLAGS=-Og -g -Wall -fcheck=all -fbacktrace"
+                " -Warray-bounds -Wunused -Wuninitialized"
+            ]
+        else:
+            args += ["-DFortran_FLAGS=-O2 -fPIC -ftree-vectorize  -fallow-argument-mismatch"]
+
+        if "+netcdf" in spec:
+            args += ["-DSIESTA_WITH_NETCDF=ON"]
+            if "+cray" in spec:
+                args += ["-DNetCDF_PATH={0}".format(os.environ.get("NETCDF_DIR", ""))]
+            else:
+                args += ["-DNetCDF_PATH={0}".format(spec["netcdf-fortran"].prefix)]
+
+        if "+elpa" in spec:
+            args += ["-DSIESTA_WITH_ELPA=ON"]
+
+        if "+mumps" in spec:
+            args += ["-DSIESTA__MUMPS=ON"]
+            args += [
+                "-DSIESTA_LINKER_FLAGS=-L{0} -lmumps_common -lzmumps".format(
+                    self.spec["mumps"].prefix.lib
+                )
+            ]
+
+        if "+metis" in spec:
+            args += ["-DSIESTA__METIS=ON"]
+            args += ["-DSIESTA_LINKER_FLAGS=-L{0} -lmetis".format(self.spec["metis"].prefix.lib)]
+
+        return args
```

Comment:
2nd part of removal of the cmake class for a later PR.
```suggestion
```

Reply:
On an HPE-Cray EX system, I successfully installed the 5.x versions without any problems using GCC@12.1 , spack> 0.21 and Cmake>3.27 .

````
==> Concretized siesta@5.0.1%gcc@12.1  +cray+debug+elpa+metis+mpi+mumps+netcdf+openmp build_system=cmake build_type=Debug ^cray-libsci+mpi+openmp
 -   ayflvth  siesta@5.0.1%gcc@12.1 cppflags='-march=znver3' +cray+debug+elpa~ipo+metis+mpi+mumps+netcdf+openmp~pexsi build_system=cmake build_type=Debug generator=make arch=linux-rhel8-zen4
[+]  r7p5ixl      ^cmake@3.27.9%gcc@12.1~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-rhel8-zen4
[+]  bo7xutf          ^curl@8.7.1%gcc@12.1~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-rhel8-zen4
[+]  kqilg34              ^nghttp2@1.57.0%gcc@12.1 build_system=autotools arch=linux-rhel8-zen4
[+]  64khgcj                  ^diffutils@3.10%gcc@12.1 build_system=autotools arch=linux-rhel8-zen4
[e]  zo6tnpu              ^openssl@1.1.1k%gcc@12.1~docs+shared build_system=generic certs=mozilla arch=linux-rhel8-zen4
[e]  qh43d7r              ^perl@5.26.3%gcc@12.1+cpanm+opcode+open+shared+threads build_system=generic patches=8cf4302 arch=linux-rhel8-zen4
[+]  ecum5ei              ^pkgconf@2.2.0%gcc@12.1 build_system=autotools arch=linux-rhel8-zen4
[+]  73yqspb          ^ncurses@6.5%gcc@12.1~symlinks+termlib abi=none build_system=autotools patches=7a351bc arch=linux-rhel8-zen4
[+]  3wvzze5          ^zlib-ng@2.1.6%gcc@12.1+compat+new_strategies+opt+pic+shared build_system=autotools arch=linux-rhel8-zen4
[e]  6g3ktqo      ^cray-libsci@23.05.1.4%gcc@12.1+mpi+openmp+shared build_system=generic arch=linux-rhel8-zen4
[e]  k7nxaro      ^cray-mpich@8.1.28%gcc@12.1+wrappers build_system=generic arch=linux-rhel8-zen4
[+]  3f7nntl      ^elpa@2023.11.001-patched%gcc@12.1~autotune~cuda~gpu_streams+mpi+openmp~rocm build_system=autotools arch=linux-rhel8-zen4
[+]  ozupbey          ^libtool@2.4.7%gcc@12.1 build_system=autotools arch=linux-rhel8-zen4
[+]  yf675er              ^findutils@4.9.0%gcc@12.1 build_system=autotools patches=440b954 arch=linux-rhel8-zen4
[+]  oszl455              ^m4@1.4.19%gcc@12.1+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-rhel8-zen4
[+]  imw5cma                  ^libsigsegv@2.14%gcc@12.1 build_system=autotools arch=linux-rhel8-zen4
[e]  hs7vfmi          ^python@3.11.5%gcc@12.1+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=13fa8bf,b0615b2,ebdca64,f2fd060 arch=linux-rhel8-zen4
[+]  gtmvflu      ^gcc-runtime@12.1%gcc@12.1 build_system=generic arch=linux-rhel8-zen4
[e]  ljqol5u      ^glibc@2.28%gcc@12.1 build_system=autotools arch=linux-rhel8-zen4
[+]  jr3usln      ^gmake@4.4.1%gcc@12.1~guile build_system=generic arch=linux-rhel8-zen4
[+]  5df5aqi      ^metis@5.1.0%gcc@12.1~gdb~int64~ipo~real64+shared build_system=cmake build_type=Release generator=make patches=4991da9,93a7903,b1225da arch=linux-rhel8-zen4
[+]  t53ux66      ^mumps@5.6.2%gcc@12.1~blr_mt+complex+double+float~incfort~int64~metis+mpi+openmp~parmetis~ptscotch~scotch+shared build_system=generic patches=373d736 arch=linux-rhel8-zen4
[e]  nwlp6rw      ^netcdf-c@4.9.0.9%gcc@12.1+blosc~byterange~dap~fsync~hdf4~jna+mpi~nczarr_zip+optimize~parallel-netcdf+pic+shared+szip+zstd build_system=autotools patches=0161eb8 arch=linux-rhel8-zen4
[e]  knndat4      ^netcdf-fortran@4.9.0.9%gcc@12.1~doc+pic+shared build_system=autotools arch=linux-rhel8-zen4



==> Installing siesta-5.0.1-ayflvthpcf26f5h2qv2pxr7u6xcmfi2n [17/17]
==> Fetching file:///opt/software/gaia/local_repo/_source-cache/archive/80/800a22a831c1d36c6f5fe4aa9c49ee510cbd49a0b2f87b3c8bf3edb6ebd0193a.tar.gz
==> No patches needed for siesta
==> siesta: Executing phase: 'cmake'
==> siesta: Executing phase: 'build'
==> siesta: Executing phase: 'install'
==> siesta: Successfully installed siesta-5.0.1-ayflvthpcf26f5h2qv2pxr7u6xcmfi2n
  Stage: 0.32s.  Cmake: 25.35s.  Build: 33.17s.  Install: 2.32s.  Post-install: 1.15s.  Total: 1m 2.81s
  ````

Could you please provide more information on the compiler used, the version of Cmake, Spack and the system on which you tried to install Siesta? if you could also provide the installation command or the concretization it would be appreciated.

Reply:
I was using Ubuntu 22.04 and did run just `spack install siesta`, that's all.

Likely the issue was that the default was `build_system=makefile`, but I've found the build_system syntax to only use `build_system=cmake` for v5, and that fixes the issue.

I'll push the updates that I made to fix this and the issues described above.

Reply:
I fixed it this morning

Reply:
Yes, it is improved in that it selects the build_system now by default, but the updated
```py
build_system(conditional("cmake", when="@5.0.0:"), "makefile", default="cmake")
```
still accepts `siesta@5 build_system=makefile` which then fails with `No such file or directory: 'Obj'`.

If you apply this change from https://github.com/CINES-HPC/spack/pull/1, spack would reject that and enforce that only build_system=cmake applies:
```py
build_system(conditional("cmake", when="@5:"), conditional("makefile", when="@:4"))
```
Also, this makes the intention and the facts that there is no choice and the fact that v4 only support `makefile` and v5 only supports `cmake` clear. Your current version lies about that.

As this is part of the other conversation, I resolve this conversation here.